altos-rust/altos-core/libs/heap/free_list_allocator/src/free_list.rs

```diff
@@ -0,0 +1,494 @@
+            if !match_condition(previous, current) {
+                previous = current;
+                current = unsafe { (*current).next_block };
+                continue;
+            }
+            break;
+        }
+        (previous, current)
+    }
+
+    // Allocate memory using the first fit strategy
+    // Returns pointer to allocated memory, or null if no memory can be found
+    pub fn allocate(&mut self, request_size: usize, request_align: usize) -> *mut u8 {
+        let using_size = alignment::align_up(request_size, self.block_hdr_size);
+        let using_align = alignment::use_align(request_align, self.block_hdr_size);
+
+        // Will return true if current size is large enough to accomodate request
+        let acceptable_block = |_, current: *mut BlockHeader| {
+            let alloc_start = alignment::align_up(current as usize, using_align);
+            let align_diff = alloc_start - current as usize;
+            let current_size = unsafe { (*current).block_size };
+
+            if current_size - align_diff < using_size { false } else { true }
+        };
+
+        let (previous, current) = self.find_block(&acceptable_block);
+        // If current is null, that means no free block was found that's large enough.
+        if current.is_null() {
+            return current as *mut u8;
+        }
+
+        let alloc_start = alignment::align_up(current as usize, using_align);
+        let align_diff = alloc_start - current as usize;
+        let current_size = unsafe { (*current).block_size };
+
+        // Block is correct size and alignment.
+        if current_size == using_size && align_diff == 0 {
+            // If at head, there is no previous to adjust
+            if self.head == current {
+                self.head = unsafe { (*self.head).next_block };
+            }
+            else {
+                unsafe { (*previous).next_block = (*current).next_block; }
+            }
+        }
+        // Current block is larger than required and has the correct alignment
+        else if current_size > using_size && align_diff == 0 {
+            unsafe { (*current).block_size -= using_size; }
+            if self.head == current {
+                self.head = self.shift_block_forward(current, using_size);
+            }
+            else {
+                unsafe {
+                    (*previous).next_block = self.shift_block_forward(current, using_size);
+                }
+            }
+        }
+        // Current block is larger than required but has exactly the right size to
+        // accomodate the alignment and size requirements
+        else if current_size == using_size + align_diff {
+            unsafe { (*current).block_size = align_diff; }
+        }
+        // Current block is larger than required and because of alignment, the allocation
+        // divides it in two.
+        else {
+            unsafe {
+                (*current).block_size = current_size - using_size - align_diff;
```

Comment:
Same as above, I think using_size and align_diff must be multiples of 8 and current_size is larger than both, so it should have at least 8 bytes remaining.